src/test/java/HttpUrlConnectionTest.java

```diff
@@ -0,0 +1,213 @@
+        fakeDataService = new FakeDataService.HttpsUrlConnection(connection);
+        RecordableHttpsURLConnection response = (RecordableHttpsURLConnection) fakeDataService.getPostsRawResponse();
+
+        // check that the replayed response contains the censored header
+        Assert.assertNotNull(response);
+        Assert.assertNotNull(response.getHeaderField("Date"));
+        String censoredHeader = response.getHeaderField("Date");
+        Assert.assertNotNull(censoredHeader);
+        Assert.assertEquals(censoredHeader, censorString);
+    }
+
+    @Test
+    public void testMatchSettings() throws Exception {
+        Cassette cassette = TestUtils.getCassette("test_match_settings");
+        cassette.erase(); // Erase cassette before recording
+
+        // record cassette with advanced settings first
+        RecordableHttpsURLConnection connection =
+                (RecordableHttpsURLConnection) HttpClients.newClient(HttpClientType.HttpsUrlConnection, FakeDataService.GET_POSTS_URL, cassette, Mode.Record);
+        FakeDataService.HttpsUrlConnection fakeDataService = new FakeDataService.HttpsUrlConnection(connection);
+        Object ignore = fakeDataService.getPostsRawResponse();
+
+        // replay cassette with default match rules, should find a match
+        connection =
+                (RecordableHttpsURLConnection) HttpClients.newClient(HttpClientType.HttpsUrlConnection, FakeDataService.GET_POSTS_URL, cassette, Mode.Replay);
+        connection.setRequestProperty("X-Custom-Header", "custom-value"); // add custom header to request, shouldn't matter when matching by default rules
+        fakeDataService = new FakeDataService.HttpsUrlConnection(connection);
+        RecordableHttpsURLConnection response = (RecordableHttpsURLConnection) fakeDataService.getPostsRawResponse();
+        Assert.assertNotNull(response);
+
+        // replay cassette with custom match rules, should not find a match because request is different (throw exception)
+        MatchRules matchRules = new MatchRules().byEverything();
+        AdvancedSettings advancedSettings = new AdvancedSettings();
+        advancedSettings.matchRules = matchRules;
+
+        connection =
+                (RecordableHttpsURLConnection) HttpClients.newClient(HttpClientType.HttpsUrlConnection, FakeDataService.GET_POSTS_URL, cassette, Mode.Replay, advancedSettings);
+        connection.setRequestProperty("X-Custom-Header", "custom-value"); // add custom header to request, causing a match failure when matching by everything
+        fakeDataService = new FakeDataService.HttpsUrlConnection(connection);
+        FakeDataService.HttpsUrlConnection finalFakeDataService = fakeDataService;
+        Assert.assertThrows(Exception.class, () -> finalFakeDataService.getPosts());
+    }
+
+    @Test
+    public void testDelay() throws Exception {
```

Comment:
I'd hope we'll quickly revisit these tests and break them out. There's way too much happening in a single test here.

Reply:
It's the exact same tests as C# :)

Reply:
Ultimately they are, there's just a lot of setup steps in each test, but few asserts